### PR TITLE
Initial SharedResizeObserver work

### PR DESCRIPTION
**Pull request by jbuckner**
This adds a SharedResizeObserver for performant resize observations.

This general use of this is to add a listener for a particular element being resized, eg:
```
// create a shared singleton of the SharedResizeObserver and pass it to any class that needs it:
this.resizeObserver = <shared instance of SharedResizeObserver>;

// get the element you want to observe
this.containerElement = document.querySelector('#container');

// add this class as an observer to the element target that you want observed
this.resizeObserver.addObserver({
  handler: this,
  target: this.containerElement,
});

// add a handleResize(entry: ResizeObserverEntry) callback
handleResize(entry: ResizeObserverEntry): void {
  const target = entry.target;
  // make sure the entry target is the one you're interested in
  if (target !== this.containerElement) return;
  // handle the resize event
  const contentRect = entry.contentRect;
  console.debug('width', contentRect.width);
}
```

### CODE REVIEW

File: test/shared-resize-observer.test.ts [hunk 1]
```diff
@@ -1,32 +1,159 @@
...
+    });
+    // trigger another handleResize, but this one won't trigger a resize event
+    el.style.width = '75px';
+    await promisedSleep(100);
+    expect(handleResizeCallCount).to.equal(2);
   });
 
-  it('passes the a11y audit', async () => {
-    const el = await fixture<SharedResizeObserver>(html`<shared-resize-observer></shared-resize-observer>`);
+  it('prevents adding a handler twice', async () => {
```
Comment (iisa) [inline test/shared-resize-observer.test.ts:129]: TY

File: src/shared-resize-observer.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import { ResizeObserver as Polyfill } from '@juggle/resize-observer';
```
Comment (iisa) [inline src/shared-resize-observer.ts:1]: yay, this will help x-browser tremendously

File: src/shared-resize-observer.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+   */
+  private resizeObserver = new ResizeObserver(entries => {
+    // This requestAnimationFrame is to throttle the refresh rate,
+    // otherwise you get a bunch of
+    // `ResizeObserver loop completed with undelivered notifications` errors
+    // The errors are not harmful, but they happen a lot, see:
+    // https://stackoverflow.com/a/58701523
+    // https://github.com/souporserious/react-measure/issues/104
+    // https://github.com/WICG/resize-observer/issues/38
+    window.requestAnimationFrame(() => {
```
Comment (iisa) [inline src/shared-resize-observer.ts:87]: good find to pipe the handlers through this. good browser support for `window.requestAnimationFrame` too

**Review by iisa — APPROVED** [review]
looks good!
core functionality is very simple yet guards against duplicate entries.

Q: is it a 1:1 match between observed element & handler?
Q: can an observed element have multiple handlers?

File: src/shared-resize-observer.ts (hunk 2 above)
Comment (jbuckner) [reply src/shared-resize-observer.ts:1]: I added this after I saw your MR come through to fix the old Safari issue 😄

**Comment by jbuckner** [comment]
> Q: is it a 1:1 match between observed element & handler?
> Q: can an observed element have multiple handlers?

An element can have multiple handlers and a handler can observe multiple elements. The [element-handler map](https://github.com/internetarchive/iaux-shared-resize-observer/pull/1/files#diff-5d778cbfed533621d4e6403918f0871f732d34709bf181c39d474350fd5f272fR107-R110) looks like this:

```
private resizeHandlers: Map<Element, Set<SharedResizeObserverResizeHandlerInterface>> = new Map();
```

So for each element, we have a `Set` of handlers. And because it's a `Set`, we won't accidentally bind the same handler to an element twice.

**Pull request merged**